### ISSUE TITLE
Llama 70B model with 2.5bpw produced wierd responses

### ISSUE THREAD

**Issue by tomleung1996** [turboderp/exllamav2#123]
Hi, I am new to LLM quantization and wanted to use text-generation-webui to run a local LLM chatbot.

ExLlamaV2 worked great when I applied it to the 13B model with 8bpw, but when it came to the 70B model with 2.5bpw, I got strange results as below:
![image](https://github.com/turboderp/exllamav2/assets/11535152/d83c6443-67de-437d-92d9-9f94101a9ff2)

I think this was not the problem of the web UI because it also occurred while using `examples/chat.py`.

**Comment by fgdfgfthgr-fox** [comment]
Are you using ROcM?

**Comment by tomleung1996** [comment]
> Are you using ROcM?

No, I am using an RTX4090 with CUDA 11.8.

**Comment by turboderp** [comment]
Which model is this, and are you sure you're using the right prompt format? From the screenshot it looks like you're running with the Llama format, which includes a BOS token at the beginning of each prompt. Some models go bonkers when they see that. TGW has the option to not include the BOS, and the CLI chatbot has a number of other prompt formats you can try (run with `-modes` to get a list of options.)

**Comment by tomleung1996** [comment]
> Which model is this, and are you sure you're using the right prompt format? From the screenshot it looks like you're running with the Llama format, which includes a BOS token at the beginning of each prompt. Some models go bonkers when they see that. TGW has the option to not include the BOS, and the CLI chatbot has a number of other prompt formats you can try (run with `-modes` to get a list of options.)

I am using `Llama2-70B-chat-hf model`.  I was using `ExLlamav2_HF` model loader, with ` Add the bos_token to the beginning of prompts ` unchecked in the parameters tab, and chatting in the chat tab using `chat-instruct` mode. However, my problem still persists.

Which `mode` should I use for the quantization model?

**Comment by turboderp** [comment]
If it's Llama2-chat, you should use the `llama` mode. There are some [conversions here](https://huggingface.co/turboderp/Llama2-70B-chat-exl2) of Llama2-70B-chat that you can compare against to see if the problem is with the quantized model or if something's going wrong in inference.

**Comment by oobabooga** [comment]
I have reproduced this issue with a `Xwin-LM-70B-V0.1-EXL2-2.500b` that I created, and the solution of not using the BOS token worked for me:

| With BOS | Without BOS |
|----------------|----------------|
| ![bos2](https://github.com/turboderp/exllamav2/assets/112222186/2be75d95-95dc-48ca-80bb-d7f90b2d3511) | ![bos](https://github.com/turboderp/exllamav2/assets/112222186/672d45e5-275a-4063-9e1c-a3ba3922788a) |

In both cases, the correct prompt format is used (`Vicuna-v1.1`).
 
Interestingly, I also created a `Llama-2-70b-chat-EXL2-2.500b` and it generated coherent outputs without removing the BOS.

For both quants, I used [this parquet file](https://huggingface.co/datasets/WizardLM/WizardLM_evol_instruct_70k/blob/refs%2Fconvert%2Fparquet/default/train/0000.parquet) and the following conversion command:

```
python convert.py -i ../Xwin-LM_Xwin-LM-70B-V0.1_safetensors -o ~/working -cf Xwin-LM-70B-V0.1-EXL2-2.500b -c Evol-Instruct-Code-80k-v1.parquet -b 2.500
```

**Comment by turboderp** [comment]
It makes sense for Llama2-chat that it would work with the BOS token, since the prompt format includes it on every round. For that Xwin model (and however many others) it'll be down to how it's been finetuned. There isn't really a "correct" way to format training examples, so as long as people publishing models don't include some examples of *fully* formatted training data (with the resulting encoding), we end up having to guess.

Here, I would speculate that the finetuning was somewhat aggressive and maybe used `<s>` in an unusual way. Maybe every training example had `<s>` followed by a newline token or some such, and now the model gets wildly confused whenever `<s>` is followed by anything else. Who can say.

To make matters worse, depending on how much the model has been finetuned (learning rate, number of epochs and so on) it may still output coherent text even if it isn't being prompted correctly, it just won't necessarily behave as intended. E.g. it might adhere to the prompt format most of the time, then suddenly deviate from it for no apparent reason.

**Comment by oobabooga** [comment]
> Here, I would speculate that the finetuning was somewhat aggressive and maybe used `<s>` in an unusual way. Maybe every training example had `<s>` followed by a newline token or some such, and now the model gets wildly confused whenever `<s>` is followed by anything else. Who can say.

That would make sense. What can be done then is to let people know about the need to remove the BOS token. I have uploaded the two quants that I mentioned to HF and added a small note to the README about this issue:

https://huggingface.co/oobabooga/Llama-2-70b-chat-EXL2-2.500b

https://huggingface.co/oobabooga/Xwin-LM-70B-V0.1-EXL2-2.500b

**Comment by tomleung1996** [comment]
> If it's Llama2-chat, you should use the `llama` mode. There are some [conversions here](https://huggingface.co/turboderp/Llama2-70B-chat-exl2) of Llama2-70B-chat that you can compare against to see if the problem is with the quantized model or if something's going wrong in inference.

The conversions you provided worked for me (without removing the BOS)! There could be something wrong in my quantization process, but I did manage to quantize a 13B model and ran it with no problem. 

I also noticed a difference that TGW can auto-detect the model loader for your conversions but not mine (even for the 13B model).

**Comment by RandomInternetPreson** [comment]
Hey All, just another data point to add to the discussion.  I quantized Xwin-LM-70B-V0.1 with 8.000 bit and it worked without needing to do anything to the default Oob text gen webui settings (BOS is being used).  I used the same model and training data as documented in Oobabooga's post, I did change the last layer bit precision to 8 instead of the default 6.

I'll be working on llama2-70B-chat conversion later today with the same 8.000 bit setup.

**Comment by greyfoxone** [comment]
> I have reproduced this issue with a `Xwin-LM-70B-V0.1-EXL2-2.500b` that I created, and the solution of not using the BOS token worked for me:
> With BOS 	Without BOS
> ![bos2](https://user-images.githubusercontent.com/112222186/277379808-2be75d95-95dc-48ca-80bb-d7f90b2d3511.png) 	![bos](https://user-images.githubusercontent.com/112222186/277379825-672d45e5-275a-4063-9e1c-a3ba3922788a.png)
> 
> In both cases, the correct prompt format is used (`Vicuna-v1.1`).
> 
> Interestingly, I also created a `Llama-2-70b-chat-EXL2-2.500b` and it generated coherent outputs without removing the BOS.
> 
> For both quants, I used [this parquet file](https://huggingface.co/datasets/WizardLM/WizardLM_evol_instruct_70k/blob/refs%2Fconvert%2Fparquet/default/train/0000.parquet) and the following conversion command:
> 
> ```
> python convert.py \
>   -i ../Xwin-LM_Xwin-LM-70B-V0.1_safetensors \
>   -o ~/working \
>   -cf Xwin-LM-70B-V0.1-EXL2-2.500b \
>   -c Evol-Instruct-Code-80k-v1.parquet \
>   -b 2.500
> ```

I have the exact same issue. I downloaded turboderp_Llama2-70B-exl2 2.5 bpw. I have a 3090 24GB. The LLM is hung up on the word "cord" mostly:
![image](https://github.com/turboderp/exllamav2/assets/18591270/7d9de023-cd9e-43a5-b20f-623c13442a4c)
How do I fix this I do not understand your comment completely. I am new to this but my understanding is that the LLM gets confused by the BOS token and that there is am option to disable it? Were do I find the option in the oobabooga gui ?

**Comment by turboderp** [comment]
It's under parameters->generation, on the right-hand side of the page.

![image](https://github.com/turboderp/exllamav2/assets/11859846/af13a2d6-9246-4128-bc89-e9aaacc50ef3)

**Comment by turboderp** [comment]
Closing this now. For reference the new measurement procedure seems to have fixed the cord cord stringbuilder issue on low bitrate models, regardless of BOS settings.
